cterasdk/core/directoryservice.py

```diff
@@ -3,14 +3,181 @@
+        :param bool,optional krb: Connect using Kerberos, defaults to `False`
+        :param list[cterasdk.core.types.ADDomainIDMapping],optional: The directory services UID/GID mapping
+        :param list[cterasdk.core.types.AccessControlEntry],optional acl: List of access control entries and their associated roles
+        :param cterasdk.core.enum.Role default: Default role if no match applies, defaults to `None`
+        :param str,optional fetch: Configure identity fetching mode, defaults to `'Lazy'`
+        """
+        param = Object()
+        param._classname = 'ActiveDirectory'  # pylint: disable=protected-access
+        param.type = directory
+        param.domain = domain
+        param.useKerberos = krb
+        param.useSSL = ssl
+        param.username = username
+        param.password = password
+        param.ou = ou
+        param.noMatchRole = default
+        param.accessControlRules = None
+        param.idMapping = None
+        param.fetchMode = fetch
+
+        if domain_controllers:
+            param.ipAddresses = Object()
+            param.ipAddresses._classname = 'DomainControlIPAddresses'  # pylint: disable=protected-access
+            param.ipAddresses.ipAddress1 = domain_controllers.primary
+            param.ipAddresses.ipAddress2 = domain_controllers.secondary
+
+        tenant = self._portal.session().user.tenant
+        logging.getLogger().info('Connecting Portal to directory services. %s', {
+            'tenant': tenant,
+            'type': type,
+            'domain': domain
+        })
+        self._connect_to_directory_services(param)
+        logging.getLogger().info('Connected Portal to directory services. %s', {'tennat': tenant, 'domain': domain})
+
+        if mapping:
+            self._configure_advanced_mapping(mapping)
+
+        if acl:
+            self._configure_access_control(acl, default)
+
+    def _connect_to_directory_services(self, param):
+        return self._portal.execute('', 'testAndSaveAD', param)
+
+    def get_advanced_mapping(self):
+        """
+        Retrieve directory services advanced mapping configuration
+
+        :returns: A dictionary of domain mapping objects
+        :rtype: dict
+        """
+        return {map.domainFlatName: map for map in self._portal.get('/directoryConnector/idMapping/map')}
+
+    def set_advanced_mapping(self, mapping):
+        """
+        Configure advanced mapping
+
+        :param list[cterasdk.core.types.ADDomainIDMapping] mapping: The directory services UID/GID mapping
+        """
+        if self._get_configuration() is None:
+            raise CTERAException('Failed to apply mapping. Not connected to directory services.')
+
+        return self._configure_advanced_mapping(mapping)
+
+    def _configure_advanced_mapping(self, mapping):
+        param = Object()
+        param._classname = 'ADIDMapping'  # pylint: disable=protected-access
+        param.map = list()
```

Comment:
Not sure I understand this code. Why can't you just assign the list?

Reply:
The "idMapping" attribute (parent object) may be None. In that scenario, the "map" attribute does not exist

Reply:
Fixed